daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala

```diff
@@ -1400,8 +1408,15 @@ object Main {
 
   private def unparseWithSAX(
     is: InputStream,
-    contentHandler: DFDL.DaffodilUnparseContentHandler): UnparseResult = {
-    val xmlReader = DaffodilSAXParserFactory().newSAXParser.getXMLReader
+    contentHandler: DFDL.DaffodilUnparseContentHandler,
+    infosetType: InfosetType.Type): UnparseResult = {
+    val xmlReader = infosetType match {
+      case InfosetType.EXI => {
+        val exiSource = new EXISource()
+        exiSource.getXMLReader
+      }
+      case _ => DaffodilSAXParserFactory().newSAXParser.getXMLReader
+    }
```

Comment:
I imagined that we would have a new `InfosetInputter` and `InfosetOutputter` specific to EXI (i.e. `EXIInfosetInputter`/`EXIInfosetOutputter`). That way anyone would be able to use this by just specifying the new infosetter without having to do all this SAX logic themselves. Is that something we're trying to avoid? That would move all of this logic out of the CLI and would make it easily usable in the API. It could stil use the SAX InfosetOutputter/Inputter under the hood, just makes it easier to use.

Another thought, this is just using our SAXInfosetOutputter and plugging that into this API--I wonder if we would get better performance if we directly use the EXI API? Converting our internal infoset representation to SAX events has some amount of overhead ([DAFFODIL-2400](https://issues.apache.org/jira/browse/DAFFODIL-2400)) that might be avoided if we call the appropriate API calls. Just skimming the code, it looks like EXIBodyEncoder is maybe one of the more important API classes and is pretty straightforward. It should map pretty closely to our InfosetOuputter API:

https://github.com/EXIficient/exificient-core/blob/master/src/main/java/com/siemens/ct/exi/core/EXIBodyEncoder.java

This also seems especially important in the case of unparsing. Daffodil SAX unparsing is extra complicated because Daffodil InfosetInputters are pull-based based, but SAX is push-based, so we have this mess of coroutines to link to two and it creates a lot of overhead. If we could directly use the EXI API, or at least a StAX based API for an EXIInfosetOutputter, I imagine performance would be significantly better.

Reply:
I'm open to ideas on this one.  @mbeckerle seemed more interested in relying on the SAX handlers.  It's probably worth it for me to do some performance testing to see how performance is with this current SAX/EXI code.  If we aren't seeing any significant performance improvements, perhaps it would be worth it to use EXI directly in a more traditional Infoset inputter/outputter.

Reply:
In addition to potential performance improvements, one benefit of creating a custom EXIInfosetInputter/Outputter is that is becomes part of the Daffodil API, and anyone using Daffodil can easily using it by just choosing the right class. With this change, EXI support is only added to the CLI--anyone wanting to using EXI must use code similar to what is in the CLI and do it themselves.

And we don't even really need a new release for that (except for maybe the reset() change?). Can we already claim we support EXI via our SAX API?